pilot/pkg/xds/ads.go

```diff
@@ -418,9 +418,21 @@ func (s *DiscoveryServer) shouldRespond(con *Connection, rejectMetric monitoring
 
 	previousInfo := con.node.Active[stype]
 
+	// If this is a case of Envoy reconnecting Istiod i.e. Istiod does not have
+	// information about this typeUrl, but Envoy sends response nonce - either
+	// because Istiod is restarted or Envoy disconnects and reconnects.
+	// We should always respond with the current resource names.
+	if previousInfo == nil {
+		adsLog.Debugf("ADS:%s: RECONNECT %s %s %s", stype, con.ConID, request.VersionInfo, request.ResponseNonce)
+		con.mu.Lock()
+		con.node.Active[stype] = &model.WatchedResource{TypeUrl: request.TypeUrl, ResourceNames: request.ResourceNames}
+		con.mu.Unlock()
+		return true
+	}
+
 	// If there is mismatch in the nonce, that is a case of expired/stale nonce.
 	// A nonce becomes stale following a newer nonce being sent to Envoy.
-	if previousInfo != nil && request.ResponseNonce != previousInfo.NonceSent {
+	if request.ResponseNonce != previousInfo.NonceSent {
 		adsLog.Debugf("ADS:%s: REQ %s Expired nonce received %s, sent %s", stype,
 			con.ConID, request.ResponseNonce, previousInfo.NonceSent)
 		xdsExpiredNonce.Increment()
```

Comment:
I think it would be safer to track not only the 'previousInfo.NonceSent' - but add a 'previousInfo.UnackedNonces' map.
This will tell us exactly if it's an out of order ACK and allow to track the latency.

I don't know what we should do if in this case the list of resources is different - since Envoy doesn't explicitly says if we should honor the 'last received' list, or if the list in the out-of-order ACK matches original transaction. 

The safest solution would be to send the union - client can ignore unexpected updates, this should be a rare case and it'll eventually recover.

Reply:
According to Harvey, if we send 2 messages envoy may not ack both and instead ack just the latest one so we need to be careful. I am starting to think we need to either invest in the xds conformance test effort and make sure we are compliant or just copy what go-control-plane I'd doing exactly. we seem to run into a lot of issues around the protocol because we do things a bit nonstandard

Reply:
IMO xds conformance test effort is better it helps to validate/fix Envoy assumptions as well